File: examples/python/actor-join/actor-join.py
# Copyright (c) 2017-2025. The SimGrid Team. All rights reserved.
#
# This program is free software you can redistribute it and/or modify it
# under the terms of the license (GNU LGPL) which comes with this package.

"""
Usage: actor-join.py platform_file [other parameters]
"""

import sys
from simgrid import Actor, Engine, Host, this_actor


def sleeper():
    this_actor.info("Sleeper started")
    this_actor.sleep_for(3)
    this_actor.info("I'm done. See you!")


def master():
    this_actor.info("Start 1st sleeper")
    actor = Host.current().add_actor("1st sleeper from master", sleeper)
    this_actor.info("Join the 1st sleeper (timeout 2)")
    actor.join(2)

    this_actor.info("Start 2nd sleeper")
    actor = Host.current().add_actor("2nd sleeper from master", sleeper)
    this_actor.info("Join the 2nd sleeper (timeout 4)")
    actor.join(4)

    this_actor.info("Start 3rd sleeper")
    actor = Host.current().add_actor("3rd sleeper from master", sleeper)
    this_actor.info("Join the 3rd sleeper (timeout 2)")
    actor.join(2)

    this_actor.info("Start 4th sleeper")
    actor = Host.current().add_actor("4th sleeper from master", sleeper)
    this_actor.info("Waiting 4")
    this_actor.sleep_for(4)
    this_actor.info("Join the 4th sleeper after its end (timeout 1)")
    actor.join(1)

    this_actor.info("Goodbye now!")

    this_actor.sleep_for(1)

    this_actor.info("Goodbye now!")


if __name__ == '__main__':
    e = Engine(sys.argv)
    if len(sys.argv) < 2:
        raise AssertionError(
            "Usage: actor-join.py platform_file [other parameters]")

    e.load_platform(sys.argv[1])

    Host.by_name("Tremblay").add_actor("master", master)

    e.run()

    this_actor.info("Simulation time {}".format(e.clock))

File: examples/c/actor-join/actor-join.c
/* Copyright (c) 2010-2020. The SimGrid Team.
 * All rights reserved.                                                     */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include "simgrid/actor.h"
#include "simgrid/engine.h"
#include "simgrid/host.h"

#include "xbt/log.h"
#include "xbt/sysdep.h"

XBT_LOG_NEW_DEFAULT_CATEGORY(actor_join, "Messages specific for this example");

static void sleeper(int argc, char* argv[])
{
  XBT_INFO("Sleeper started");
  sg_actor_sleep_for(3);
  XBT_INFO("I'm done. See you!");
}

static void master(int argc, char* argv[])
{
  sg_actor_t actor;

  XBT_INFO("Start sleeper");
  actor = sg_actor_create("sleeper from master", sg_host_self(), sleeper, 0, NULL);
  XBT_INFO("Join the sleeper (timeout 2)");
  sg_actor_join(actor, 2);

  XBT_INFO("Start sleeper");
  actor = sg_actor_create("sleeper from master", sg_host_self(), sleeper, 0, NULL);
  XBT_INFO("Join the sleeper (timeout 4)");
  sg_actor_join(actor, 4);

  XBT_INFO("Start sleeper");
  actor = sg_actor_create("sleeper from master", sg_host_self(), sleeper, 0, NULL);
  XBT_INFO("Join the sleeper (timeout 2)");
  sg_actor_join(actor, 2);

  XBT_INFO("Start sleeper");
  actor = sg_actor_create("sleeper from master", sg_host_self(), sleeper, 0, NULL);
  sg_actor_ref(actor); // We have to take that ref because the actor will stop before we join it
  XBT_INFO("Waiting 4");
  sg_actor_sleep_for(4);
  XBT_INFO("Join the sleeper after its end (timeout 1)");
  sg_actor_join(actor, 1);
  sg_actor_unref(actor); // Avoid to leak memory

  XBT_INFO("Goodbye now!");

  sg_actor_sleep_for(1);

  XBT_INFO("Goodbye now!");
}

int main(int argc, char* argv[])
{
  simgrid_init(&argc, argv);
  xbt_assert(argc == 2, "Usage: %s platform_file\n\tExample: %s msg_platform.xml\n", argv[0], argv[0]);

  simgrid_load_platform(argv[1]);

  sg_actor_create("master", sg_host_by_name("Tremblay"), master, 0, NULL);

  simgrid_run();

  XBT_INFO("Simulation time %g", simgrid_get_clock());

  return 0;
}
